🚦 Change severity of the error messages

rowanc1 opened the pull request:
This adds a `error_rules` to the project configuration, which can be used to disable logging rules in the CLI.

fwkoch commented on packages/myst-config/src/errorRules/validators.ts:
```diff
@@ -0,0 +1,63 @@
+  const id = validateString(value.id, incrementOptions('id', opts));
+  if (!id) return undefined;
+  const output: ErrorRule = { id };
+  if (defined(value.severity)) {
```
Feels a little ambiguous about what happens when only `id` is specified. Looks like it remains `undefined` - then, in `addWarningForFile` the default will be replaced by `undefined` and hit the `default` case in the switch statement... Hmm, then in the store, we will also get an undefined warning `kind`.

I think we need to make `severity` required on the type, and set a default here during validation?

fwkoch replied:
That begs the question what the default is... If I put `error_rule: my-id` in my config... am I expecting that will suppress it...? Based on the name "error rule" I'd more expect the default behaviour to be "Error on rule my-id" - so the default `severity` is `'error'`?

But then what's the use-case? Are we intending this for error suppression as the default behaviour? If so, maybe we need to rethink `error_rule`...?

fwkoch commented on packages/myst-config/src/errorRules/validators.ts:
```diff
@@ -0,0 +1,63 @@
+    }) as ErrorRule['severity'];
+    if (choice) output.severity = choice;
+  }
+  // Put the unknown fields back on the object
```
My understanding is these unknown fields are intended to differentiate between multiple errors of the same `ruleId`? This feels maybe too flexible? To filter on these different entries, we are going to need to expand the line in `addWarningForFile`:
```typescript
const handler = config?.error_rules?.find((rule) => rule.id === opts.ruleId);
```
to something more like
```typescript
const handler = config?.error_rules?.find((rule) => {
  return rule.id === opts.ruleId && someOtherComparison(rule, opts);
});
```
If we allow arbitrary data, that `someOtherComparison` function becomes complicated and potentially rule-id-dependent. If instead we just have, say, `key`, then it's just an equality check. (Of course, we will still need rule-id-dependent key generation, but that can happen where the rule is actually getting violated, and by the time we get here, it's just a string.)

That also makes user feedback in the logging easier: "To suppress errors of this type, add `my-rule-id` to `error_rules`; to suppress this error only, add `key: my-rule-key`"

fwkoch replied:
Being a little stricter on this would allow:

```yaml
error_rules:
  - rule: doi-check
    severity: ignore
    keys:
      - bibKey1
      - bibKey2
      - bibKey3
```

to coerce to:

```yaml
error_rules:
  - rule: doi-check
    severity: ignore
    key: bibKey1
  - rule: doi-check
    severity: ignore
    key: bibKey2
  - rule: doi-check
    severity: ignore
    key: bibKey3
```

Which I think is nice? Easy for the user, easy for the handling logic.

fwkoch commented on packages/myst-config/tests/examples.spec.ts:
```diff
@@ -0,0 +1,77 @@
+    return tests;
+  });
+
+casesList.forEach(({ title, kind, cases }) => {
```
Oh brother, no nice yaml tests in this package yet... 😆 Thanks for adding!

fwkoch commented on packages/myst-cli/src/utils/addWarningForFile.ts:
```diff
@@ -26,7 +27,18 @@ export function addWarningForFile(
+      severity = handler.severity as WarningKind;
+    }
+  }
+  switch (severity) {
```
After this switch statement, we need to add
```typescript
session.log.debug(`To suppress this message, add "${rule.id}" to "error_rules" in your project config`)
```

fwkoch replied:
This make me think about my previous comment about default severity. It's easier to say "add X to error_rules" than to say "add an entry to error_rules with id X and severity info" - so maybe that means the default is `severity: info`. In that case, maybe it should be, like `error_exceptions`?

```yaml
error_exceptions:
  - my-id
```

sounds like we are excepting that error. But

```yaml
error_exceptions:
  - id: my-id
    severity: warn
```

also still makes sense - it's an exception because we are changing the severity.

fwkoch replied:
Also... maybe we don't want to tell the user how to suppress the message. That might make it too easy to abuse...? _But_ we definitely need to debug log the `ruleId`; that way, if they read the docs, they have all the info they need to suppress the message.

If we don't log the `ruleId` this whole feature feels inaccessible to users.

fwkoch commented on packages/myst-config/src/errorRules/validators.ts:
```diff
@@ -0,0 +1,63 @@
+  if (defined(value.severity)) {
+    const choice = validateChoice(value.severity, {
+      ...incrementOptions('severity', opts),
+      choices: ['ignore', 'warn', 'error'],
```
🤔 Would be nice if `validateChoice` had an option for `caseInsensitive`... (out of scope of this PR!)

fwkoch reviewed (changes requested):
This is great! There's a couple necessary changes I think (log `ruleId`s to users; make sure `kind: undefined` doesn't make it to the warnings in the session store), as well as a bunch of other thoughts I had after marinating on this for a while.

rowanc1 commented:
@fwkoch thanks for the pair-programming to go through this. We addressed all the comments in the PR now I think, I have also gone through and updated all failing links.

fwkoch commented on docs/settings.md:
````diff
@@ -52,3 +52,22 @@ beamer
+        - https://flaky-connection.com
+```
+
+The `severity` of each rule can be set to `ignore`, `warn`, or `error`. If the rule is triggered, then the severity listed will be adopted rather than the default log message severity. The default severity for rules included in the list is `ignore`, which means that simply listing the rule IDs as strings will ignore those rules. To discover the rule ID, run myst in debug mode to get the error (an ...
````
Updates to these docs are great 👍
